tests/unit/shared-heap/shared_heap_test.cc

```diff
@@ -92,149 +94,956 @@ destroy_module_env(struct ret_env module_env)
-        const char *s = wasm_runtime_get_exception(tmp_module_env.wasm_module_inst);
-        printf("exception: %s\n", s);
-        goto test_failed;
+        const char *s =
+            wasm_runtime_get_exception(tmp_module_env.wasm_module_inst);
+        ADD_FAILURE() << "Failed to wasm_runtime_call_wasm with "
+                      << "exception: " << s;
     }
 
+fail2:
     wasm_runtime_detach_shared_heap(tmp_module_env.wasm_module_inst);
+fail1:
     destroy_module_env(tmp_module_env);
+fail0:
     return;
-test_failed:
-    destroy_module_env(tmp_module_env);
-    EXPECT_EQ(1, 0);
 }
 
 TEST_F(shared_heap_test, test_shared_heap_basic)
 {
-    SharedHeapInitArgs args;
+    SharedHeapInitArgs args = { 0 };
     WASMSharedHeap *shared_heap = nullptr;
     uint32 argv[1] = { 0 };
 
     args.size = 1024;
     shared_heap = wasm_runtime_create_shared_heap(&args);
 
     if (!shared_heap) {
-        printf("Failed to create shared heap\n");
-        EXPECT_EQ(1, 0);
+        FAIL() << "Failed to create shared heap";
     }
 
-    // test wasm
-    test_shared_heap(shared_heap, "test.wasm", "test", 1, argv);
+    test_shared_heap(shared_heap, "test.wasm", "test", 0, argv);
     EXPECT_EQ(10, argv[0]);
 
-    // test aot
-    test_shared_heap(shared_heap, "test.aot", "test", 1, argv);
+    test_shared_heap(shared_heap, "test.aot", "test", 0, argv);
     EXPECT_EQ(10, argv[0]);
 
+    test_shared_heap(shared_heap, "test_chain.aot", "test", 0, argv);
+    EXPECT_EQ(10, argv[0]);
 }
 
 TEST_F(shared_heap_test, test_shared_heap_malloc_fail)
 {
-    SharedHeapInitArgs args;
+    SharedHeapInitArgs args = { 0 };
     WASMSharedHeap *shared_heap = nullptr;
     uint32 argv[1] = { 0 };
 
     args.size = 1024;
     shared_heap = wasm_runtime_create_shared_heap(&args);
 
     if (!shared_heap) {
-        printf("Failed to create shared heap\n");
-        EXPECT_EQ(1, 0);
+        FAIL() << "Failed to create shared heap";
     }
 
-    // test wasm
-    test_shared_heap(shared_heap, "test.wasm", "test_malloc_fail", 1, argv);
+    test_shared_heap(shared_heap, "test.wasm", "test_malloc_fail", 0, argv);
     EXPECT_EQ(1, argv[0]);
 
-    // test aot
-    test_shared_heap(shared_heap, "test.aot", "test_malloc_fail", 1, argv);
+    test_shared_heap(shared_heap, "test.aot", "test_malloc_fail", 0, argv);
     EXPECT_EQ(1, argv[0]);
+
+    test_shared_heap(shared_heap, "test_chain.aot", "test_malloc_fail", 0,
+                     argv);
+    EXPECT_EQ(1, argv[0]);
+}
+
+TEST_F(shared_heap_test, test_preallocated_shared_heap_malloc_fail)
+{
+    SharedHeapInitArgs args = { 0 };
+    WASMSharedHeap *shared_heap = nullptr;
+    uint32 argv[1] = { 0 }, BUF_SIZE = os_getpagesize();
+    uint8 preallocated_buf[BUF_SIZE];
+
+    /* create a preallocated shared heap */
+    args.pre_allocated_addr = preallocated_buf;
+    args.size = BUF_SIZE;
+    shared_heap = wasm_runtime_create_shared_heap(&args);
+    if (!shared_heap) {
```

Comment:
Who is going to release this shared_heap at the end of every case(if the case create_test_shared_heap()?

Reply:
runtime will destroy the shared heap list and free the resources